### PR TITLE
SUBMARINE-256. Interpreter support cluster mode

### PR DESCRIPTION
**Pull request by liuxunorg**
### What is this PR for?
After the interpreter integrates the cluster module,
When the interpreter is started, the `interpreter ID`, `IP` and `port` of the interpreter are registered in the metadata of the cluster, so that the interpreter can be found in the entire cluster and used by the notebook and workbench server.

### What type of PR is it?
[Feature]

### Todos
* [ ] - Task

### What is the Jira issue?
* https://issues.apache.org/jira/browse/SUBMARINE-256

### How should this be tested?
* [CI Pass](https://travis-ci.org/liuxunorg/hadoop-submarine/builds/601138608)
* InterpreterClusterTest.java

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

File: submarine-commons/commons-cluster/src/main/java/org/apache/submarine/commons/cluster/ClusterClient.java [hunk 1]
```diff
@@ -25,23 +25,23 @@
...
 
-  private static ClusterManagerClient instance = null;
+  private static ClusterClient instance = null;
 
   // Do not use the getInstance function in the test case,
   // which will result in an inability to update the instance according to the configuration.
-  public static ClusterManagerClient getInstance() {
-    synchronized (ClusterManagerClient.class) {
+  public static ClusterClient getInstance() {
+    synchronized (ClusterClient.class) {
```
Comment (yuanzac) [inline submarine-commons/commons-cluster/src/main/java/org/apache/submarine/commons/cluster/ClusterClient.java]: Can we change it like this:
```
if (instance == null) {
  synchronized (ClusterClient.class) {
    if (instance == null) {
        instance = new ClusterClient();
      }
    }
  }
return instance;
```

File: submarine-commons/commons-cluster/src/main/java/org/apache/submarine/commons/cluster/ClusterServer.java [hunk 2]
```diff
@@ -58,26 +58,26 @@
...
+  private ClusterServer() {
     super();
   }
 
   // Do not use the getInstance function in the test case,
   // which will result in an inability to update the instance according to the configuration.
-  public static ClusterManagerServer getInstance() {
-    synchronized (ClusterManagerServer.class) {
+  public static ClusterServer getInstance() {
+    synchronized (ClusterServer.class) {
```
Comment (yuanzac) [inline submarine-commons/commons-cluster/src/main/java/org/apache/submarine/commons/cluster/ClusterServer.java]: Can we change it like cluster client?

File: submarine-workbench/interpreter/interpreter-engine/pom.xml [hunk 3]
```diff
@@ -36,6 +36,18 @@
   <description>Submarine Interpreter Engine</description>
 
   <dependencies>
+    <dependency>
+      <groupId>org.apache.submarine</groupId>
+      <artifactId>commons-utils</artifactId>
+      <version>0.3.0-SNAPSHOT</version>
```
Comment (yuanzac) [inline submarine-workbench/interpreter/interpreter-engine/pom.xml]: Can we change 0.3.0-SNAPSHOT to ${submarine.version}?

File: submarine-commons/commons-cluster/src/main/java/org/apache/submarine/commons/cluster/ClusterClient.java (hunk 1 above)
Comment (liuxunorg) [reply submarine-commons/commons-cluster/src/main/java/org/apache/submarine/commons/cluster/ClusterClient.java]: Done.

File: submarine-commons/commons-cluster/src/main/java/org/apache/submarine/commons/cluster/ClusterServer.java (hunk 2 above)
Comment (liuxunorg) [reply submarine-commons/commons-cluster/src/main/java/org/apache/submarine/commons/cluster/ClusterServer.java]: Done.

File: submarine-workbench/interpreter/interpreter-engine/pom.xml (hunk 3 above)
Comment (liuxunorg) [reply submarine-workbench/interpreter/interpreter-engine/pom.xml]: Done.

**Comment by yuanzac** [comment]
Hi @liuxunorg,
The CI failed. Can you help to take a look

**Comment by liuxunorg** [comment]
CI Pass: https://travis-ci.org/liuxunorg/hadoop-submarine/builds/601338814

**Pull request closed without merging**